encodings/sbcs-codec.js

```diff
@@ -1,72 +1,77 @@
-    }
+// correspond to encoded bytes (if 128 - then lower half is ASCII).
 
-    this.decodeBuf = Buffer.from(codecOptions.chars, 'ucs2');
-    
-    // Encoding buffer.
-    var encodeBuf = Buffer.alloc(65536, iconv.defaultCharSingleByte.charCodeAt(0));
+exports._sbcs = class SBCSCodec {
+    constructor(codecOptions, iconv) {
+        if (!codecOptions)
+            throw new Error("SBCS codec is called without the data.")
 
-    for (var i = 0; i < codecOptions.chars.length; i++)
-        encodeBuf[codecOptions.chars.charCodeAt(i)] = i;
+        // Prepare char buffer for decoding.
+        if (!codecOptions.chars || (codecOptions.chars.length !== 128 && codecOptions.chars.length !== 256))
+            throw new Error("Encoding '"+codecOptions.type+"' has incorrect 'chars' (must be of len 128 or 256)");
 
-    this.encodeBuf = encodeBuf;
-}
+        if (codecOptions.chars.length === 128) {
+            var asciiString = "";
+            for (let i = 0; i < 128; i++)
+                asciiString += String.fromCharCode(i);
+            codecOptions.chars = asciiString + codecOptions.chars;
+        }
 
-SBCSCodec.prototype.encoder = SBCSEncoder;
-SBCSCodec.prototype.decoder = SBCSDecoder;
+        const decodeBuf = new Uint16Array(codecOptions.chars.length);
 
+        for (let i = 0; i < codecOptions.chars.length; i++)
+            decodeBuf[i] = codecOptions.chars.charCodeAt(i);
 
-function SBCSEncoder(options, codec) {
-    this.encodeBuf = codec.encodeBuf;
-}
+        this.decodeBuf = decodeBuf;
 
-SBCSEncoder.prototype.write = function(str) {
-    var buf = Buffer.alloc(str.length);
-    for (var i = 0; i < str.length; i++)
-        buf[i] = this.encodeBuf[str.charCodeAt(i)];
-    
-    return buf;
-}
+        // Encoding buffer.
+        const encodeBuf = iconv.backend.allocBytes(65536, iconv.defaultCharSingleByte.charCodeAt(0));
 
-SBCSEncoder.prototype.end = function() {
+        for (let i = 0; i < codecOptions.chars.length; i++)
+            encodeBuf[codecOptions.chars.charCodeAt(i)] = i;
+
+        this.encodeBuf = encodeBuf;
+    }
+    get encoder() { return SBCSEncoder; }
+    get decoder() { return SBCSDecoder; }
 }
 
+class SBCSEncoder {
+    constructor(opts, codec, backend) {
+        this.backend = backend;
+        this.encodeBuf = codec.encodeBuf;
+    }
 
-function SBCSDecoder(options, codec) {
-    this.decodeBuf = codec.decodeBuf;
-}
+    write(str) {
+        const bytes = this.backend.allocBytes(str.length);
 
-SBCSDecoder.prototype.write = function(buf) {
-    // Strings are immutable in JS -> we use ucs2 buffer to speed up computations.
-    var decodeBuf = this.decodeBuf;
-    var newBuf = Buffer.alloc(buf.length*2);
-    var idx1 = 0, idx2 = 0;
-    for (var i = 0; i < buf.length; i++) {
-        idx1 = buf[i]*2; idx2 = i*2;
-        newBuf[idx2] = decodeBuf[idx1];
-        newBuf[idx2+1] = decodeBuf[idx1+1];
+        for (let i = 0; i < str.length; i++)
+            bytes[i] = this.encodeBuf[str.charCodeAt(i)];
```

Comment:
love how concise that is

Reply:
The funny part is, I was expecting the the whole migration will be more like a manual labor. Just changing calls to `Buffer.from` and `Buffer.toString` to our custom functions. So it took me some time to figure out how change the code. But in the end was interesting to dig a little bit deeper into encodings business.